File: packages/cli-js/src/utils/context.ts
import { relative, resolve } from "path";
import fs from "fs-extra";
import { getDefaultContextDirs } from "./paths.js";

async function resolveContextBases(cwd: string, contextDir?: string): Promise<string[]> {
  if (contextDir) {
    const base = resolve(cwd, contextDir);
    if (!(await fs.pathExists(base))) return [];
    const stat = await fs.stat(base);
    if (!stat.isDirectory()) {
      throw new Error(`Context path is not a directory: ${contextDir}`);
    }
    return [base];
  }

  const bases: string[] = [];
  for (const candidate of getDefaultContextDirs(cwd)) {
    if (!(await fs.pathExists(candidate))) continue;
    const stat = await fs.stat(candidate);
    if (!stat.isDirectory()) continue;
    bases.push(candidate);
  }
  return bases;
}

export async function loadContextDir(cwd: string, contextDir?: string): Promise<string | undefined> {
  const root = resolve(cwd);
  const bases = await resolveContextBases(root, contextDir);
  if (bases.length === 0) return undefined;

  const mdPaths: string[] = [];
  for (const base of bases) {
    const stack = [base];
    while (stack.length > 0) {
      const dir = stack.pop() as string;
      const entries = await fs.readdir(dir, { withFileTypes: true });
      for (const entry of entries) {
        const next = resolve(dir, entry.name);
        if (entry.isDirectory()) {
          stack.push(next);
        } else if (entry.isFile() && entry.name.toLowerCase().endsWith(".md")) {
          mdPaths.push(next);
        }
      }
    }
  }

  const withMtime: { path: string; mtime: number }[] = [];
  for (const path of mdPaths) {
    const fileStat = await fs.stat(path);
    withMtime.push({ path, mtime: fileStat.mtimeMs });
  }
  withMtime.sort((a, b) => b.mtime - a.mtime || a.path.localeCompare(b.path));

  const blocks: string[] = [];
  for (const item of withMtime) {
    const content = (await fs.readFile(item.path, "utf-8")).trim();
    if (!content) continue;
    blocks.push(`### ${relative(root, item.path).replace(/\\/g, "/")}\n\n${content}`);
  }
  if (blocks.length === 0) return undefined;
  return blocks.join("\n\n");
}

export interface MergeContextOpts {
  contextDir?: string;
  inlineContext?: string;
}

export async function loadMergedContext(cwd: string, opts: MergeContextOpts): Promise<string | undefined> {
  const parts: string[] = [];
  const dirContent = await loadContextDir(cwd, opts.contextDir);
  if (dirContent) parts.push(dirContent);

  if (opts.inlineContext?.trim()) parts.push(opts.inlineContext.trim());
  if (parts.length === 0) return undefined;
  return parts.join("\n\n");
}

File: packages/cli-py/planforge/utils/context.py
"""Conversation context loaders (markdown directory and inline text)."""

from pathlib import Path

from planforge.utils.paths import get_default_context_dirs


def _resolve_context_bases(cwd: str, context_dir: str | None) -> list[Path]:
    if context_dir:
        base = (Path(cwd) / context_dir).resolve()
        if not base.exists():
            return []
        if not base.is_dir():
            raise OSError(f"Context path is not a directory: {context_dir}")
        return [base]

    bases: list[Path] = []
    for candidate in get_default_context_dirs(cwd):
        base = Path(candidate).resolve()
        if base.exists() and base.is_dir():
            bases.append(base)
    return bases


def load_context_dir(cwd: str, context_dir: str | None) -> str | None:
    root = Path(cwd).resolve()
    bases = _resolve_context_bases(cwd, context_dir)
    if not bases:
        return None

    files: list[Path] = []
    for base in bases:
        files.extend(path for path in base.rglob("*") if path.is_file() and path.name.lower().endswith(".md"))
    files.sort(key=lambda path: (-path.stat().st_mtime, path.as_posix()))

    blocks: list[str] = []
    for path in files:
        content = path.read_text(encoding="utf-8").strip()
        if not content:
            continue
        rel = path.relative_to(root).as_posix()
        blocks.append(f"### {rel}\n\n{content}")
    if not blocks:
        return None
    return "\n\n".join(blocks)


def load_merged_context(
    cwd: str,
    *,
    context_dir: str | None,
    inline_context: str | None,
) -> str | None:
    parts: list[str] = []

    dir_content = load_context_dir(cwd, context_dir)
    if dir_content:
        parts.append(dir_content)

    if (inline_context or "").strip():
        parts.append((inline_context or "").strip())

    return "\n\n".join(parts) if parts else None
